### PR TITLE
Introduction documentation for tensor mechanics

### PR DESCRIPTION
**Pull request by sapitts**
### Documentation for Tensor Mechanics

In this commit I've created a separate folder for the introduction/overview materials we had on the mooseframework.com tensor mechanics page, quite possibly undermining all of the work @aeslaughter did to join the src code and documentation.  If the `Introduction` folder goes against the philosophy of this new documentation system, I'll split the information out into the separate classes.

@WilkAndy, @jiangwen84, @dschwen, and @tonkmr I've pulled documentation that each of you wrote on the website and 'repackaged' that documentation here.  I've tried to keep the Introduction pages shorter, under the assumption that we would keep the more in-depth documentation in the individual classes.  Please let me know where I've made a mistake in the documentation / where my copy and pasting skills have left rough transitions.
### Test Cases

I changed a few of the names of the materials used in the input file (`[./name]`) to create a unique combination to include specific input file blocks in this documentation.

@aeslaughter please review.
Refs #6699

### CODE REVIEW

**Comment by aeslaughter** [comment]
@sapitts I will take a look at this after we get #7778 merged, which should fix your input file include problems.

**Comment by dschwen** [comment]
👍  Thanks for doing this!

File: docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergence1DRSpherical.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergence1DRSpherical.md]: Breaking up the file name with spaces helps to not overrun the blue left sidebar block

File: docs/content/modules/tensor_mechanics/introduction/StressDivergence.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/introduction/StressDivergence.md]: @aeslaughter In this line I would like to link to the StressDivergenceTensors.md file, but the webpage links to StressDivergenceTensorsTruss.md instead--is there a way to search for only the provided string instead of files that contain the provided string?

File: docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergenceRSphericalTensors.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergenceRSphericalTensors.md]: This ability to link to a separate file to duplicate content is quite handy.  I made a separate folder for these *.md files that are common to multiple material/kernel descriptions at the same level as `tensor_mechanics/systems` but did not include the folder in the `pages.yml` file.  Is this separate folder okay?  My goal was to avoid completely overfilling the Materials folder

File: docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md]: @milljm Here's that really long figure caption I promised as a test case for the css template.  Alternatively I can strive to be less verbose

File: docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md]: @aeslaughter I was hoping to use this include command to list the child classes which inherit from this class but did not have any luck.  Is there something in the yml file I need to add as with the tests?

File: docs/content/modules/tensor_mechanics/introduction/StressDivergence.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/introduction/StressDivergence.md]: The "auto" operates based on MOOSE syntax, so if there is not syntax for StressDivergenceTensors then the auto will never find it. It does do partial matches such as you have found here. I suspect that this is what is happening.

I think that I need to re-work this a bit to just take .md filenames and make the proper link.

File: docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergenceRSphericalTensors.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergenceRSphericalTensors.md]: You can do what you want, there is no getting around having a lot of content. We have a lot of code that needs to be documented, it is unavoidable. However, I would prefer if you would rename this directory to "common" or something in lower-case. Camel case items generally mirror the input syntax. This isn't a rule, put perhaps is should be just to help keep md files associated directly with C++ objects easier to identify. Just an idea.

File: docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md]: No, we certainly want verbose captions.

**Comment by aeslaughter** [comment]
@sapitts This is looking excellent, thanks for helping us work out the kinks in the system.

**Comment by sapitts** [comment]
I'll get the file name fixed first thing tomorrow. Thank you for the help!

File: docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md]: There is a way, you need to edit the 'moosedocs.yml' and add the directory you want to search.

File: docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergenceRSphericalTensors.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergenceRSphericalTensors.md]: I've changed the file name structure in my latest commit to `common_documents/supplementalStressDivergenceKernels.md` Does this help with the identification of the C++ object associated md files?

File: docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md]: Would it be possible to link under just tensor mechanics in `moosedocs.yml` for the tensor mechanics child classes?  I'm hesitant to add all of the module classes to the top level `links` in moosedoc to avoid creating monster length child class lists in framework.  Of course this is a fussy request that can wait.

File: docs/bib/tensor_mechanics.bib
Comment (aeslaughter) [inline docs/bib/tensor_mechanics.bib]: Could you also move the "moose.bib" file into this folder.

File: docs/content/modules/tensor_mechanics/introduction/Overview.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/introduction/Overview.md]: Didn't know we could do this, it looks nice but we may need to reduce the font-size of the quote. It is bigger than the regular text, what do you think? This is something we need to set in the docs/css/moose.css file, not really relevant to this PR.

File: docs/content/modules/tensor_mechanics/introduction/VisualizingTensors.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/introduction/VisualizingTensors.md]: It is possible to link to an existing input file here, rather than copy paste? I know that this syntax isn't likely to change, but you never no.

File: docs/content/modules/tensor_mechanics/introduction/VisualizingTensors.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/introduction/VisualizingTensors.md]: Same comment as above, is there an input file somewhere with this information.

File: docs/content/modules/tensor_mechanics/systems/AuxKernels/RadialDisplacementCylinderAux.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/AuxKernels/RadialDisplacementCylinderAux.md]: Can you add some content here?

File: docs/content/modules/tensor_mechanics/systems/AuxKernels/RadialDisplacementSphereAux.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/AuxKernels/RadialDisplacementSphereAux.md]: Can you add some information about this class?

File: docs/content/modules/tensor_mechanics/systems/Bounds/RankTwoScalarAux.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/Bounds/RankTwoScalarAux.md]: Why didn't you want this again?

File: docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergence1DRSpherical.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergence1DRSpherical.md]: I would leave the long name for now, we will be improving this sidebar in the coming weeks. I am hoping we can handle the long names. So, it is better to leave it for now so that it matches the object name.

File: docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergenceRSphericalTensors.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/Kernels/StressDivergenceRSphericalTensors.md]: Like above and the others below, please just leave the long name for now

File: docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md]: We talked about this, I prefer a monster list rather than an incomplete list. The list should be scrollable, so the length is handled.

File: docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md
Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md]: 👍

**Review by aeslaughter — CHANGES_REQUESTED** [review]

**Comment by aeslaughter** [comment]
You have one image that looks kind of flow chart like, you might consider changing it to use graphiz: http://mooseframework.com/docs/moose_docs/site/content/utilities/documentation/moose_flavored_markdown/#flow-charts

**Comment by sapitts** [comment]
@dschwen Would you take another look at this PR for content please?  I took most of this content from the mooseframework wiki, but it'd be good to have you look over this material to ensure I didn't make mistakes in explaining the theory

File: docs/content/modules/tensor_mechanics/systems/Bounds/RankTwoScalarAux.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Bounds/RankTwoScalarAux.md]: Bounds was an identical copy of the Auxkernels folder, so I removed bounds since auxkernels matches the code structure

File: docs/content/modules/tensor_mechanics/introduction/VisualizingTensors.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/introduction/VisualizingTensors.md]: I could manufacture an input file with this information; this sort of input file would fit most logically with another PR I have open to update the elasticity tensor.  May I wait to switch out this file with the links to the input file in another PR?

Comment (aeslaughter) [inline docs/content/modules/tensor_mechanics/introduction/VisualizingTensors.md]: Fine with me.

**Review by aeslaughter — APPROVED** [review]

**Comment by aeslaughter** [comment]
I think that this documentation looks great, I will leave it to @dschwen to make the finial review of the actual content. @sapitts Thanks for getting this going and working through the bugs.

File: docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/Materials/RecomputeRadialReturn.md]: This particular class is not used in any input files, so I've deleted the input file link in just this file.  The other classes are showing the input files nicely though

File: docs/content/modules/tensor_mechanics/systems/AuxKernels/RadialDisplacementCylinderAux.md
Comment (sapitts) [inline docs/content/modules/tensor_mechanics/systems/AuxKernels/RadialDisplacementCylinderAux.md]: These classes were checked in by Ben after I started working on the documentation, so these stubs were automatically generated for those new classes.  I'll talk with Ben about these auxkernels and add documentation in a new PR.

File: docs/bib/tensor_mechanics.bib
Comment (sapitts) [inline docs/bib/tensor_mechanics.bib]: done

**Comment by sapitts** [comment]
I've addressed @aeslaughter's comments; @dschwen, in all of your spare time would you take another look over this commit for theoretical content please?  The content has not changed much (just the Radial Recompute stuff added) since you last gave this a thumbs up last week--most of my recent commit amends have been to fix formatting.

**Review by dschwen — APPROVED** [review]
Ok, I did a single pass over this and it looks fine. This should be merged as is and then iterated on if necessary.

**Comment by sapitts** [comment]
Thank you both!

**Comment by WilkAndy** [comment]
Hi Stephanie, How does this work?  I want to write some new documentation, but don't know how to "latex" my text, or where it's supposed to display to (a website?  my computer?)

**Comment by sapitts** [comment]
Hi Andy, thanks for helping to write documentation!  @aeslaughter  and @milljm have written some instructions on how to set up your computer to generate the documentation and instructions on how to use the Moose flavored markdown that they have created here: http://mooseframework.org/docs/moose_docs/site/content/utilities/documentation/overview/ . There's also talk over in the Moose cube about another package with support for the documentation system being on the way--Jason knows much more about this new package than me.

If you are editing `*.md` files that already exist in tensor mechanics (in `moose/docs/content/modules/tensor_mechanics/systems/` ), you can always simply add those files to your commit, activate the Documentation test under Additional Recipes to Run in civet, and then moose build will helpfully comment with a temporary website including your new documentation changes.

**Comment by WilkAndy** [comment]
Thanks @sapitts for your great explanation.

**Comment by aeslaughter** [comment]
Andy, We are still working on improving the system and I would say we are
in the beta testing phase right now. So, don't go too crazy at this point
as things will be changing often and without warning.

**Comment by WilkAndy** [comment]
OK, thanks,

a

**Comment by WilkAndy** [comment]
I didn't find the instructions on that website totally clear, but i struggled on, hoping for the best.  Now i get:

```
[wil04q][~/projects/moose/docs]> ./moosedocs.py -h
Traceback (most recent call last):
  File "./moosedocs.py", line 16, in <module>
    import MooseDocs
  File "/Users/wil04q/projects/moose/python/MooseDocs/__init__.py", line 7, in <module>
    import extensions
  File "/Users/wil04q/projects/moose/python/MooseDocs/extensions/__init__.py", line 1, in <module>
    from MooseMarkdown import MooseMarkdown
  File "/Users/wil04q/projects/moose/python/MooseDocs/extensions/MooseMarkdown.py", line 4, in <module>
    import markdown
ImportError: No module named markdown
```

What have i done wrong?

**Comment by WilkAndy** [comment]
I'm possibly getting somewhere with this, but the instructions are just a bit unclear for a copy-and-paste guy like me.  Here are my comments on the instruction page:

```
http://mooseframework.org/docs/moose_docs/site/content/utilities/documentation/setup/

"Follow the getting started instructions on this page..."  On *what* page?  Does this mean the current webpage (.../documentation/setup) or the git-lfs.github.com page?  I had a look at the latter and it's got some "getting started" instructions, but i don't know what they mean.

"git lfs pull"  From which directory do i run this?  did it from projects/moose.

"Manual Setup"  "If you are not using a MOOSE package, then the following steps may..."  Should this "may" actually be "must"?  Because i AM using a MOOSE package, i skipped this step, and continued through to trying to run moosedocs.py, but then i got
ImportError: No module named markdown
So i went back and followed the "Manual Setup" instructions, because i saw that something called "markdown" was mentioned.
I'm not sure what "source activate docs" does.  Do i have to do this every time i write new documentation?


"Configuration"  I didn't understand this at all.  I can find python/MooseDocs/extensions/MooseMarkdown.py, but i don't know what i have to do with it, nor what this docs/moosedocs.yml is.


I continued to look at the instructions, and it seemed like "moosedocs.py" was a pretty fundamental thing.  But i can't run it.  After doing the "Manual Setup" (above) i now get

  File "/Users/wil04q/projects/moose/python/MooseDocs/extensions/MooseBibtex.py", line 6, in <module>
    from pybtex.plugin import find_plugin
ImportError: No module named pybtex.plugin

```

**Comment by aeslaughter** [comment]
Andy,

There is a reason why this site is not public yet, these instructions are
written for when it is public and the instructions are still changing as we
move forward.

On Wed, Oct 19, 2016 at 6:17 PM, Andy Wilkins <notifications@github.com>
wrote:

> I'm possibly getting somewhere with this, but the instructions are just a
> bit unclear for a copy-and-paste guy like me. Here are my comments on the
> instruction page:
>
> http://mooseframework.org/docs/moose_docs/site/content/utilities/documentation/setup/
>
> "Follow the getting started instructions on this page..."  On *what* page?  Does this mean the current webpage (.../documentation/setup) or the git-lfs.github.com page?  I had a look at the latter and it's got some "getting started" instructions, but i don't know what they mean.
>
> git-lfs.github.com

>
> "git lfs pull"  From which directory do i run this?  did it from projects/moose.
>
> Anywhere inside the repository should work.

>
> "Manual Setup"  "If you are not using a MOOSE package, then the following steps may..."  Should this "may" actually be "must"?  Because i AM using a MOOSE package, i skipped this step, and continued through to trying to run moosedocs.py, but then i got
> ImportError: No module named markdown
> So i went back and followed the "Manual Setup" instructions, because i saw that something called "markdown" was mentioned.
> I'm not sure what "source activate docs" does.  Do i have to do this every time i write new documentation?
>
> We don't have a package containing these dependencies yet because they are
still changing. "source activate doc", activates a python environment
suitable for the documentation, it allows you to have an environment for
working in without messing up the regular python stuff for MOOSE. If you
want to run 'moosedocs.py' then you need to have this environment activated.

>
>
> "Configuration"  I didn't understand this at all.  I can find python/MooseDocs/extensions/MooseMarkdown.py, but i don't know what i have to do with it, nor what this docs/moosedocs.yml is.
>
>  This is the beginning of how to setup documentation in an application
that does not have it, this is still changing and hopefully getting simpler
soon. And, eventually this stuff will come with stork. I am still working
on this part heavily. If you are documenting in MOOSE you probably don't
need to worry about this too much.

Where are you trying to add documentation?

>
> I continued to look at the instructions, and it seemed like "moosedocs.py" was a pretty fundamental thing.  But i can't run it.  After doing the "Manual Setup" (above) i now get
>
>   File "/Users/wil04q/projects/moose/python/MooseDocs/extensions/MooseBibtex.py", line 6, in <module>
>     from pybtex.plugin import find_plugin
> ImportError: No module named pybtex.plugin
>
> There is a missing dependency in the instructions, just run "pip install
pybtex" after activating the docs environment.

**Comment by sapitts** [comment]
My apologies Andrew and Andy, I jumped the gun on sharing the site before
it was ready to share publicly

**Comment by WilkAndy** [comment]
No, i think you did the right thing by telling me, thanks @sapitts .  I "wasted" only a small amount of time and then gave up, and i understand this is still in "alpha" so it's completely fine for things not to work perfectly.  This looks like a major addition to MOOSE that will really improve it from a user's perspective.

I think i understand what to do now with my doco (just documenting the equations behind a plasticity class i'm currently building - when i say "just", so far the doco is 13 pages long!)

My only question at present is: how do i do the latex equivalent of /label{my.equation} and /ref{my.equation} in the markdown?

**Comment by WilkAndy** [comment]
Oh, this is for a TensorMechanics class, by the way - in moose/modules.

**Comment by aeslaughter** [comment]
On Thu, Oct 20, 2016 at 1:56 PM, Andy Wilkins <notifications@github.com>
wrote:

> No, i think you did the right thing by telling me, thanks @sapitts
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_sapitts&d=CwMFaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=h7heP8xwI1i_HikChvhFbEBurKirgfOCdwgBxB9lM8c&m=_f15Dq1ol1UwMX01mJBm1D7GhBImBimUvSfFpIT8wQM&s=__iRwlz7dx3JLOsELQpZ275QNayIC6W3IoC7tAd7KJ4&e=>
> . I "wasted" only a small amount of time and then gave up, and i understand
> this is still in "alpha" so it's completely fine for things not to work
> perfectly. This looks like a major addition to MOOSE that will really
> improve it from a user's perspective.
>
> I think i understand what to do now with my doco (just documenting the
> equations behind a plasticity class i'm currently building - when i say
> "just", so far the doco is 13 pages long!)
>
> My only question at present is: how do i do the latex equivalent of
> /label{my.equation} and /ref{my.equation} in the markdown?
>
Currently this is not supported, I plan to add it but I am not sure what it
will look like at this point, but I am hoping to support this soonish.

**Comment by WilkAndy** [comment]
Aha, thanks for planning to implement the "label" + "ref" stuff, @aeslaughter .  I've written a latex version of my doco, but will wait till you've done the implementation before i make my .md file.  In the case at hand, the doco is quite lengthy and makes a lot of use of labels.

**Pull request merged**
